GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs

```diff
@@ -157,16 +136,16 @@ private bool FindBranch([NotNull] Node node)
             /// First/second branch does not matter because it is the message of the current node.
             /// </summary>
             /// <param name="node">the node of the revision to evaluate</param>
-            /// <param name="descJunction">
-            /// the descending junction the node is part of
-            /// (used for the decision whether the node belongs the first or second branch of the merge)
+            /// <param name="parent">
+            /// the node's parent in the branch which is currently descended
+            /// (used for the decision whether the node belongs to the first or second branch of the merge)
             /// </param>
-            private bool CheckForMerge([NotNull] Node node, [CanBeNull] Junction descJunction)
+            private bool CheckForMerge([NotNull] RevisionGraphRevision node, [CanBeNull] RevisionGraphRevision parent)
             {
-                bool isTheFirstBranch = descJunction == null || node.Ancestors.Count == 0 || node.Ancestors.First() == descJunction;
+                bool isTheFirstBranch = parent == null || node.Parents.Count == 0 || node.Parents.First() == parent;
```

Comment:
The sequence of Parents is not "1st destination branch, 2nd merged branch(es)" for the optimized graph anymore.
It depends on the attempt to reduce the crossing of branches now. That's why my approach does not work anymore.
@spdr870, is the information which is the destination branch of a merge, stored somewhere?
If not, could you give me a hint where to add it?

Reply:
I was looping at the code and noticed you use the stattsegments and endsegments collections. Those are unordered. You probably do not want to use those. You should only use the parents and child collections. The parents collection is ordered. The first parent is the branch where the other parents are merged into. I do not change the parent order,  it is I  the same order as git returned it.

I'd you want I could fake a closer look at your code.

Reply:
Ignore my last commit... The parents are not ordered. They are stored in a concurrentbag. The startsegments are ordered. If you need the parents to be ordered, you can probably change the type ConcurrentBag into SynchonizedCollection.

Reply:
In tests I did earlier I noticed that in many situations the SynchronizedCollection is even faster then the ConcurrentBag. Only in situations where many threads are used, the concurrentbag is faster. In our case, there usually is 1 thread and exceptionally max 2.

The EndSegments and StartSegments are mainly meant to be used for the graph itself. For your usage, I prefer only to use the Parents and Children. I might even remove the StartSegments from the revision class, since it is not needed there. It can easily be created later.